storage/storage-interface/src/finality_view.rs

```diff
@@ -0,0 +1,159 @@
+    pub fn new(reader: Db) -> Self {
+        Self {
+            reader,
+            finalized_ledger_info: RwLock::new(None),
+        }
+    }
+}
+
+impl<Db: DbReader> FinalityView<Db> {
+    /// Updates the information on the latest finalized block at the specified height.
+    pub fn set_finalized_block_height(&self, height: u64) -> Result<()> {
+        let (_start_ver, end_ver, block_event) = self.reader.get_block_info_by_height(height)?;
+        let block_hash = block_event.hash()?;
+        let block_info = BlockInfo::new(
+            block_event.epoch(),
+            block_event.round(),
+            block_hash,
+            self.reader.get_accumulator_root_hash(end_ver)?,
+            end_ver,
+            block_event.proposed_time(),
+            None,
+        );
+        // Sanity checks: finalization should not be set on an empty database,
+        // the finality version should not exceed the latest committed.
+        match self.reader.get_latest_state_checkpoint_version()? {
+            None => return Err(AptosDbError::Other("no ledger states to finalize".into())),
+            Some(ver) => {
+                if end_ver > ver {
+                    return Err(AptosDbError::Other(format!(
+                        "finality version {end_ver} exceeds committed version {ver}"
+                    )));
+                }
+            },
+        }
+        let ledger_info = LedgerInfo::new(block_info, HashValue::zero()); // we don't use consensus_data_hash
+        let aggregate_signature = aggregate_signature::AggregateSignature::empty(); // we don't use
+                                                                                    // aggregate_signatures
+        let ledger_info = LedgerInfoWithSignatures::new(ledger_info, aggregate_signature);
+        let mut fin_legder_info = self.finalized_ledger_info.write().unwrap();
+        *fin_legder_info = Some(ledger_info);
+        Ok(())
+    }
+}
+
+impl<Db: DbReader> DbReader for FinalityView<Db> {
+    fn get_read_delegatee(&self) -> &dyn DbReader {
+        &self.reader
+    }
+
+    fn get_latest_ledger_info_option(&self) -> Result<Option<LedgerInfoWithSignatures>> {
+        let fin_ledger_info = self.finalized_ledger_info.read().unwrap();
+        Ok(fin_ledger_info.clone())
+    }
+
+    fn get_latest_version(&self) -> Result<Version> {
+        let fin_ledger_info = self.finalized_ledger_info.read().unwrap();
+        fin_ledger_info
+            .as_ref()
+            .map(|li| li.ledger_info().version())
+            .ok_or_else(|| AptosDbError::NotFound("finalized version".into()))
+    }
+
+    fn get_latest_state_checkpoint_version(&self) -> Result<Option<Version>> {
+        let fin_ledger_info = self.finalized_ledger_info.read().unwrap();
+        let version = fin_ledger_info
+            .as_ref()
+            .map(|li| li.ledger_info().version());
+        Ok(version)
+    }
+
+    // TODO: override any other methods needed to maintain the illusion.
```

Comment:
What are these methods ATM?

Reply:
Anything in the API that does not use `get_latest_ledger_info` or construct a state view around the latest version, but takes a version as a parameter, would be able to get objects from non-finalized state versions. If this is an issue, we need to override the relevant methods and error out on versions that are newer than the finalized one.